url9512d8fa89_bdkosher_card_games_tgz-pJ8-org_washcom_cardgames_battleroyale_BattleRoyaleGameSimulatorJ8/src/org/washcom/util/Deck.java

```diff
@@ -0,0 +1,269 @@
+     * the first card is the top card of the deck and so on.
+     * 
+     * @param nbrOfTs - must be greater than 0
+     * @return
+     * @throws DeckExhaustedException - if there aren't enough cards in the deck which can be drawn; the
+     *      thrown exception will contain the partial list of drawn cards.
+     */
+    public List<T> draw(int nbrOfTs) throws DeckExhaustedException {
+        return draw(nbrOfTs, true);
+    }
+
+    /**
+     * Remove the bottom card from the deck. 
+     * 
+     * @return
+     * @throws DeckExhaustedException - if the deck is empty
+     */
+    public T drawFromBottom() throws DeckExhaustedException {
+        return draw(false);
+    }
+
+    /**
+     * Draw the given number of cards from the bottom of the deck. The order is preserved in the returned list:
+     * the first card is the bottom card of the deck and so on.
+     * 
+     * @param nbrOfTs - must be greater than 0
+     * @return
+     * @throws DeckExhaustedException - if there aren't enough cards in the deck which can be drawn; the
+     *      thrown exception will contain the partial list of drawn cards.
+     */
+    public List<T> drawFromBottom(int nbrOfTs) throws DeckExhaustedException {
+        return draw(nbrOfTs, false);
+    }
+
+    /**
+     * Remove a card from the deck. If {@code top} is true, it's the top card; otherwise, it's the bottom card.
+     *
+     * @param top - true if to be drawn from the top; false if to be drawn from the bottom 
+     * @return
+     * @throws DeckExhaustedException - if the deck is empty
+     */
+    private T draw(boolean top) throws DeckExhaustedException {
+        if (internal.isEmpty()) {
+            throw new DeckExhaustedException();
+        }
+        return top ? internal.removeFirst() : internal.removeLast();
+    }
+
+    /**
+     * Draw the given number of cards from the top of the deck (if {@code top} is true) or the bottom of the deck
+     * (if {@code top} is false). The order is preserved in the returned list.
+     * 
+     * @param nbrOfTs - must be non-negative
+     * @param top - true if to be drawn from the top; false if to be drawn from the bottom
+     * @return
+     * @throws DeckExhaustedException - if there aren't enough cards in the deck which can be drawn; the
+     *      thrown exception will contain the partial list of drawn cards.
+     */
+    private List<T> draw(int nbrOfTs, boolean top) throws DeckExhaustedException {
+        if (nbrOfTs < 0) {
+            throw new IllegalArgumentException("Integer arg must be 1 or more.");
+        }
+        if (nbrOfTs == 0) {
+            return Collections.emptyList();
+        }
+        List<T> hand = new ArrayList<>();
+        try {
+            for (int i = nbrOfTs; i > 0; --i) {
+                hand.add(top ? internal.removeFirst() : internal.removeLast());
+            }
+        } catch (NoSuchElementException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)